### PR TITLE
Enable Persisting Experiment Definitions To Disk

### PR DESCRIPTION
**Pull request by bonedaddy**
Almost done, opening the WIP PR to get some initial feedback

# Purpose

Enable persisting experiment definitions to disk. The addition of this into `master`brings #39  2/3 of the way to being finished, with the remaining 1/3 being the creation of the experiments API

# Changes

* Add test for `metadata/DB`
* Add tests for experiment definition persisting
* Modify `metadata/DB` to work with the new version of `ExperimentDefinition`
* When creating a new DB, create the root directory if it does not exist

### CODE REVIEW

File: experiments/experiments_test.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+				t.Fatal("bad status")
+			}
+			if !reflect.DeepEqual(exp1.Definition, exp2.Definition) {
+				t.Fatal("bad definition")
+			}
+			exp3, err := db.GetExperiment(ctx, exp1.ID)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if exp1.ID != exp3.ID {
```
Comment (hinshun) [inline experiments/experiments_test.go:53]: `require.Equal(t, exp1.ID, exp3.ID)`

File: metadata/db.go [hunk 2]
```diff
@@ -139,6 +140,13 @@ type db struct {
 }
 
 func NewDB(root string) (DB, error) {
+	if _, err := os.Stat(root); os.IsNotExist(err) {
+		if err := os.Mkdir(root, os.FileMode(0740)); err != nil {
```
Comment (hinshun) [inline metadata/db.go:144]: Curious why the odd permission? Typical directories are `0775`.

File: metadata/experiment.go [hunk 3]
```diff
@@ -254,34 +266,19 @@ func readExperiment(bkt *bolt.Bucket, experiment *Experiment) error {
 	})
 }
 
-func readExperimentDefinition(bkt *bolt.Bucket) (ExperimentDefinition, error) {
+func readExperimentDefinition(bkt *bolt.Bucket, experiment *Experiment) (ExperimentDefinition, error) {
 	var edef ExperimentDefinition
-	/* TODO: enable in a followup PR
 	dbkt := bkt.Bucket(bucketKeyDefinition)
 	if dbkt == nil {
+		// TODO: should we error here
```
Comment (hinshun) [inline metadata/experiment.go:273]: No, validation is performed at write time. If our schema evolves, we want to read out what we can without erroring.

File: experiments/experiments_test.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+	"github.com/google/uuid"
+)
+
+func TestExperimentDefinition(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	db, cleanup := newTestDB(t, "exptestdir")
+	defer func() {
+		if err := cleanup(); err != nil {
+			t.Fatal(err)
```
[the remark this answers is not in the document]
Comment (bonedaddy) [reply experiments/experiments_test.go:22]: yea

File: experiments/experiments_test.go (hunk 1 above)
Comment (bonedaddy) [reply experiments/experiments_test.go:52]: will do

File: metadata/db.go (hunk 2 above)
Comment (bonedaddy) [reply metadata/db.go:144]: No real reason, guess it's just a habit.

File: metadata/experiment.go (hunk 3 above)
Comment (bonedaddy) [reply metadata/experiment.go:273]: makes sense

File: experiments/experiments_test.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+			"../cue/cue.mod/p2plab_example2.cue",
+		}
+		for _, sourceFile := range sourceFiles {
+			name := strings.Split(sourceFile, "/")
+			exp1 := newTestExperiment(t, sourceFile, name[len(name)-1])
+			ids = append(ids, exp1.ID)
+			exp2, err := db.CreateExperiment(ctx, exp1)
+			if err != nil {
+				t.Fatal(err)
+			}
```
Comment (hinshun) [inline experiments/experiments_test.go:38]: Still needs `require.NoError(t, err)`.

File: experiments/experiments_test.go [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+			prevUpdateAt := exp.UpdatedAt
+			exp.Labels = append(exp.Labels, "test label")
+			exp, err = db.UpdateExperiment(ctx, exp)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if exp.UpdatedAt.Before(prevUpdateAt) {
```
Comment (hinshun) [inline experiments/experiments_test.go:74]: `require.True(t, exp.UpdatedAt.Before(prevUpdatedAt))`

File: experiments/experiments_test.go [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+			ctx,
+			ids,
+			[]string{"should be present"},
+			[]string{"test label"},
+		)
+		if err != nil {
+			t.Fatal(err)
+		}
+		for _, exp := range exps {
+			if len(exp.Labels) != 1 {
```
Comment (hinshun) [inline experiments/experiments_test.go:90]: `require.Len(t, exp.Labels, 1)`

File: experiments/experiments_test.go (hunk 5 above)
Comment (bonedaddy) [reply experiments/experiments_test.go:38]: ah shoot my bad, i will apply the requires to everything else

**Comment by bonedaddy** [comment]
I made sure to replace all manual error checks, and other checks with `require.XXX` unfortunately there was one bit that I couldn't use `require.XXX` (checking if the ID is 1 of 2 available IDs) so I left that as is, but swapped to `t.Error` instead

File: experiments/experiments_test.go (hunk 6 above)
Comment (bonedaddy) [reply experiments/experiments_test.go]: I changed this to instead of check `exp.UpdateAt.Before` to examine `exp.UpdateAt.After`

**Pull request merged**